ocs_ci/ocs/provider_client_upgrade.py

```diff
@@ -0,0 +1,227 @@
+            csv_name_pre_upgrade=self.csv_name_pre_upgrade,
+        ):
+            try:
+                if sample:
+                    log.info("Upgrade success!")
+                    break
+            except TimeoutException:
+                raise TimeoutException("No new CSV found after upgrade!")
+
+    def collect_data(self, pod_name_pattern):
+        """
+        Collect operators installed on provider related pods and csv data
+        """
+        pod_data = pod.get_all_pods(namespace=self.namespace)
+        self.pod_name_pattern = pod_name_pattern
+        for p in pod_data:
+            if self.pod_name_pattern in p.get()["metadata"]["name"]:
+                pod_obj = OCP(
+                    namespace=self.namespace,
+                    resource_name=p.get()["metadata"]["name"],
+                    kind="Pod",
+                )
+                if self.upgrade_phase == "pre_upgrade":
+                    self.pre_upgrade_data["age"] = pod_obj.get_resource(
+                        resource_name=p.get()["metadata"]["name"], column="AGE"
+                    )
+                    self.pre_upgrade_data["pod_status"] = pod_obj.get_resource_status(
+                        resource_name=p.get()["metadata"]["name"]
+                    )
+                if self.upgrade_phase == "post_upgrade":
+                    self.post_upgrade_data["age"] = pod_obj.get_resource(
+                        resource_name=p.get()["metadata"]["name"], column="AGE"
+                    )
+                    self.post_upgrade_data["pod_status"] = pod_obj.get_resource_status(
+                        resource_name=p.get()["metadata"]["name"]
+                    )
+
+        # get pre-upgrade csv
+        csv_objs = CSV(namespace=self.namespace)
+        for csv in csv_objs.get()["items"]:
+            if self.operator_name in csv["metadata"]["name"]:
+                csv_obj = CSV(
+                    namespace=self.namespace, resource_name=csv["metadata"]["name"]
+                )
+                if self.upgrade_phase == "pre_upgrade":
+                    self.pre_upgrade_data["version"] = csv_obj.get_resource(
+                        resource_name=csv_obj.resource_name, column="VERSION"
+                    )
+                    try:
+                        self.pre_upgrade_data["version"]
+                    except KeyError:
+                        log.error(
+                            f"Couldn't capture Pre-upgrade CSV version for {self.operator_name}"
+                        )
+                if self.upgrade_phase == "post_upgrade":
+                    if self.upgrade_version in csv["metadata"]["name"]:
+                        self.post_upgrade_data["version"] = csv_obj.get_resource(
+                            resource_name=csv_obj.resource_name, column="VERSION"
+                        )
+                    try:
+                        self.post_upgrade_data["version"]
+                    except KeyError:
```

Comment:
catching this exception may leave a hole in upgrade logic. we probably should try parse version with other approaches